### PR TITLE
Feature/group search

### PR DESCRIPTION
**Pull request by eladex**
* create ES repositories for persons & organizationGroups
* refactor ES client and index settings
* refactor query extraction from the request
* add feature: search inside a group path

### CODE REVIEW

File: src/group/organizationGroup/organizationGroup.textSearch.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { IOrganizationGroup } from './organizationGroup.interface';
+
+export type GroupQuery = {
+  nameAndHierarchyTerms: string | string[];
+};
+
+export type GroupFilters = {
+  hierarchyPath: string;
+};
```
Comment (eladex) [inline src/group/organizationGroup/organizationGroup.textSearch.ts:9]: add `isAlive` filter

File: src/group/organizationGroup/organizationGroup.elasticSearchRepository.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        esb.termsQuery(f, value) : 
+        esb.termQuery(f, value)
+    );
+    const queryBody = esb.requestBodySearch().query(
+      esb.boolQuery()
+      .should(should)
+      .filter(filter)
+      .minimumShouldMatch(1)
+    ).toJSON();
+    console.log(JSON.stringify(queryBody));
```
Comment (eladex) [inline src/group/organizationGroup/organizationGroup.elasticSearchRepository.ts:43]: delete

File: src/group/organizationGroup/organizationGroup.extractQuery.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { GroupFilters } from './organizationGroup.controller';
+import Query from '../../types/Query';
+import { extract, replaceValues } from '../../utils';
+import singleValue from '../../helpers/makeSingleValue';
+
+
+const filterKeys: (keyof GroupFilters)[] = ['hierarchyPath'];
```
Comment (eladex) [inline src/group/organizationGroup/organizationGroup.extractQuery.ts:7]: add isAlive

File: src/person/person.elasticSearchRepository.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+type PersonSource = IPerson & {
+  hierarchyPath: string;
+};
+
+const { indexNames: { persons: indexName } } = config.elasticSearch;
+
+class PersonElasticSearchRepository extends ElasticSearchBaseRepository<IPerson> implements PersonTextSearch {
+
+  constructor(elasticClient?: Client, queryConfig?: QueryConfig) {
```
Comment (eladex) [inline src/person/person.elasticSearchRepository.ts:17]: accept indexName parameter

File: src/person/person.elasticSearchRepository.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import { config } from '../config/config';
+import { DomainSeperator, domainMap } from '../utils';
+
+type PersonSource = IPerson & {
+  hierarchyPath: string;
+};
+
+const { indexNames: { persons: indexName } } = config.elasticSearch;
+
+class PersonElasticSearchRepository extends ElasticSearchBaseRepository<IPerson> implements PersonTextSearch {
```
Comment (eladex) [inline src/person/person.elasticSearchRepository.ts:15]: shoud be `ElasticSearchBaseRepository<PersonSource >`

File: src/person/person.elasticSearchRepository.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+  async searchByQuery(query: PersonSearchQuery) {
+    const body = QueryBuilder.buildBoolQuery(query, {
+      fullName: {
+        context: FieldContext.Query,
+        fuzzy: true,
+      },
+    });
+    console.log(JSON.stringify(query));
+    console.log(JSON.stringify(body));
```
Comment (eladex) [inline src/person/person.elasticSearchRepository.ts:29]: delete

File: src/elasticsearch/elasticSearchBaseRepository.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    return [];
+  }
+
+  public async findById(id: any) {
+    const res = await this._client.get({
+      id,
+      index: this._indexName,
+    });
+    return res.body._source as T;
+  }
```
Comment (eladex) [inline src/elasticsearch/elasticSearchBaseRepository.ts:87]: solve type issues

File: src/person/person.extractQuery.ts [hunk 8]
```diff
@@ -20,27 +22,39 @@ export const queryParamsRenameMap = {
   responsibilitylocation: 'responsibilityLocation',
   mobilephone: 'mobilePhone',
   fullname: 'fullName',
-};
+  rank: 'rank',
+  status: 'status',
+  responsibility: 'responsibility',
+  job: 'job',
```
Comment (eladex) [inline src/person/person.extractQuery.ts:28]: remove this object, not needed

File: src/person/person.extractQuery.ts [hunk 9]
```diff
@@ -20,27 +22,39 @@ export const queryParamsRenameMap = {
...
+} as const;
 
-export const filterQueryAllowedFields = ['currentUnit', 'domainUsers.dataSource', 'entityType', 
-  'job', 'rank', 'responsibility', 'serviceType', 'status'];
+const { aliases: { persons: personAliases }, defaults: { persons: personDefaults } } = config.queries;
 
-export const searchQueryAllowedFields = filterQueryAllowedFields.concat(['fullName']);
+const filterFields: (keyof PersonFilter)[] = ['currentUnit', 'domainUsers.dataSource', 
+  'entityType', 'job', 'rank', 'responsibility', 'serviceType', 'status', 'hierarchyPath'];
+const searchFields: (keyof PersonSearchQuery)[] = [...filterFields, 'fullName'];
```
Comment (eladex) [inline src/person/person.extractQuery.ts:35]: keep name of variables consistent with the group.extractQuery

File: src/person/person.route.ts [hunk 10]
```diff
@@ -1,29 +1,28 @@
...
 
 persons.use('/', AuthMiddleware.verifyToken, PermissionMiddleware.hasBasicPermission);
 
-persons.get('/', ch(Person.getPersons, (req: Request) => [extractFilterQueryFields(req.query)]));
+persons.get('/', ch(Person.getPersons, (req: Request) => [extractFilterQuery(req.query)]));
 
 persons.get('/search', ch(Person.searchPersons, 
-  (req: Request) => [extractSearchQueryFields(req.query)])
+  (req: Request) => [extractSearchQuery(req.query)])
 );
```
Comment (eladex) [inline src/person/person.route.ts:18]: it extracts the `hierarchyPath` which is not supported in this route

File: src/person/person.route.ts [hunk 11]
```diff
@@ -1,29 +1,28 @@
...
 import { atCreateFieldCheck, atUpdateFieldCheck } from './person.route.validator';
-import { extractFilterQueryFields, extractSearchQueryFields } from './person.extractQuery';
+import { extractFilterQuery, extractSearchQuery } from './person.extractQuery';
 
 const persons = Router();
 
 persons.use('/', AuthMiddleware.verifyToken, PermissionMiddleware.hasBasicPermission);
 
-persons.get('/', ch(Person.getPersons, (req: Request) => [extractFilterQueryFields(req.query)]));
+persons.get('/', ch(Person.getPersons, (req: Request) => [extractFilterQuery(req.query)]));
```
Comment (eladex) [inline src/person/person.route.ts:14]: it extracts the `hierarchyPath` which is not supported in this route

Comment (eladex) [reply src/person/person.route.ts:14]: maybe use the TextSearch for all queries that may return many results

File: src/person/person.route.ts [hunk 12]
```diff
@@ -1,29 +1,28 @@
...
+  (req: Request) => [extractSearchQuery(req.query)])
 );
 
 persons.get('/getUpdated/:from',  
   validatorMiddleware(Vld.dateOrInt, ['from'], 'params'), 
   ch(Person.getUpdatedFrom, (req: Request) => {
     let from = req.params.from;
     if (typeof(from) === 'number') from = new Date(from);
-    return [from, new Date(), extractFilterQueryFields(req.query)];
+    return [from, new Date(), extractFilterQuery(req.query)];
```
Comment (eladex) [inline src/person/person.route.ts:25]: it extracts the `hierarchyPath` which is not supported in this route

File: src/person/person.extractQuery.ts [hunk 13]
```diff
@@ -20,27 +22,39 @@ export const queryParamsRenameMap = {
...
+  const { fullName, ...filters } = extract(query, searchFields);
+  const filtersWithAliases = extractFilterQuery(filters);
+  return {
+    ...filtersWithAliases,
+    fullName: singleValue(fullName),
+  };
+};
+
+const applyDefaultsAndAliases = (query: Query<Partial<PersonFilter>>): Query<Partial<PersonFilter>> => {
+  const withDefaults = { ...personDefaults, ...query } as Query<Partial<PersonFilter>>;
```
Comment (eladex) [inline src/person/person.extractQuery.ts:58]: the 'as' conversion may not be needed

File: src/person/person.route.ts (hunk 11 above)
Comment (eladex) [reply src/person/person.route.ts:14]: need to check if ES limits the result size

File: src/person/person.textSearch.interface.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+  entityType: string | string[];
+  'domainUsers.dataSource': string | string[];
+  rank: string | string[];
+  responsibility: string | string[];
+  hierarchyPath: string;
+  fullName: string;
+}>;
+
+export interface PersonTextSearch {
+  searchByQuery(query: PersonSearchQuery): Promise<IPerson[]>;
```
Comment (eladex) [inline src/person/person.textSearch.interface.ts:14]: maybe refactor to `searchByFullName(fullName, filters)` and make the `fullName` required

File: src/group/organizationGroup/organizationGroup.textSearch.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+export interface OrganizationGroupTextSearch {
+  searchByNameAndHierarchy(query: Partial<GroupQuery>, filters?: Partial<GroupFilters>)
+    : Promise<IOrganizationGroup[]>;
+}
```
Comment (eladex) [inline src/group/organizationGroup/organizationGroup.textSearch.ts:14]: maybe refactor to `searchByNameAndHierarchy(nameAndHierarchyTerms, filter)` and make `nameAndHierarchyTerms` required

**Review by eladex — COMMENTED** [review]
* update the docs!!
* fix tests

**Comment by yonatandt** [comment]
@eladex are you reviewing your own code???

File: docker-compose-config/monstache/config/scripts/transform_group.js [hunk 16]
```diff
@@ -1,3 +1,3 @@
 module.exports = function(doc) {
-  return _.omit(doc, 'children', 'ancestors', 'isALeaf', 'hierarchy');
+  return doc;
```
Comment (yonatandt) [inline docker-compose-config/monstache/config/scripts/transform_group.js:2]: looks like a very complicated function... 😕 
You should at least add some lines of documentation here.

File: docker-compose-config/monstache/config/scripts/transform_person.js [hunk 17]
```diff
@@ -7,5 +7,9 @@ module.exports = function(doc) {
     lastName = '';
 
   doc.fullName = doc.firstName + lastName;
+  
+  var hierarchyPath = doc.hierarchy.join('/');
```
Comment (yonatandt) [inline docker-compose-config/monstache/config/scripts/transform_person.js:11]: `var`??? 😱

File: src/group/organizationGroup/organizationGroup.controller.ts [hunk 18]
```diff
@@ -6,10 +6,22 @@ import { PersonRepository } from '../../person/person.repository';
...
+
+export type GroupQuery = {
+  name: string,
+  hierarchy: string;
+};
 
 export class OrganizationGroup {
   static _organizationGroupRepository: OrganizationGroupRepository = new OrganizationGroupRepository();
   static _personRepository: PersonRepository = new PersonRepository();
+  static _organizationGroupTextSearch: OrganizationGroupTextSearch = organizationGroupElasticRepository;
```
Comment (yonatandt) [inline src/group/organizationGroup/organizationGroup.controller.ts:24]: Is this a public or a private member variable? 🤔 
An underscore usually indicates a private member variable, but in typescript all members of a class are public by default.

File: docker-compose-config/monstache/config/scripts/transform_group.js (hunk 16 above)
Comment (eladex) [reply docker-compose-config/monstache/config/scripts/transform_group.js:2]: you are probably right

File: src/elasticsearch/elasticSearchBaseRepository.ts (hunk 7 above)
Comment (eladex) [reply src/elasticsearch/elasticSearchBaseRepository.ts:87]: deleted override

File: src/person/person.route.ts (hunk 11 above)
Comment (eladex) [reply src/person/person.route.ts]: decided to exclude `hierarchyPath` from the object to send to the controller

File: src/person/person.route.ts (hunk 12 above)
Comment (eladex) [reply src/person/person.route.ts]: decided to exclude hierarchyPath from the object to send to the controller

File: docker-compose-config/monstache/config/scripts/transform_person.js (hunk 17 above)
Comment (eladex) [reply docker-compose-config/monstache/config/scripts/transform_person.js:11]: monstache js plugin  doesn't support ES6

File: src/group/organizationGroup/organizationGroup.controller.ts [hunk 19]
```diff
@@ -240,6 +253,14 @@ export class OrganizationGroup {
     return res;
   }
 
+  static async searchGroups(query: Partial<GroupQuery>) {
+    const { hierarchy, name, ...filters } = query;
+    // the query makes sense only if any of these fields requested
+    if (!hierarchy && !name) return [];
+    return OrganizationGroup._organizationGroupTextSearch
+      .searchByNameAndHierarchy({ name, hierarchy }, filters);
```
Comment (eladex) [inline src/group/organizationGroup/organizationGroup.controller.ts:261]: may call `searchByNameAndHierarchy` with object that have empty values (on existing keys)

File: src/group/organizationGroup/organizationGroup.extractQuery.ts [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+import { extract, isBooleanString, pickSingleValue } from '../../utils';
+import { config } from '../../config/config';
+
+
+type GroupSearchQuery = GroupQueryDTO & { nameAndHierarchy: string };
+
+const { organizationGroups: defaults } = config.queries.defaults;
+
+const extractFilterKeys: (keyof GroupFilters)[] = ['isAlive'];
+const extractKeys: (keyof GroupSearchQuery)[] = ['hierarchy', 'name', 'nameAndHierarchy', 'isAlive'];
```
Comment (eladex) [inline src/group/organizationGroup/organizationGroup.extractQuery.ts:12]: `[...extractFilterKeys, name, ......]`

File: src/group/organizationGroup/organizationGroup.extractQuery.ts [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+
+
+function extractFilters(query: Query<Partial<GroupFilters>>): Partial<GroupFilters> {
+  const { isAlive } = applyDefultsAndAliases(extract(query, extractFilterKeys));
+  return {
+    ...!!isAlive && {
+      isAlive: pickSingleValue(isAlive) === 'true',
```
Comment (eladex) [inline src/group/organizationGroup/organizationGroup.extractQuery.ts:19]: `toLowerCase()`

File: src/group/organizationGroup/organizationGroup.extractQuery.ts [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+      hierarchy: pickSingleValue(hierarchy),
+    },
+    ...!!name && {
+      name: pickSingleValue(name),
+    },
+    ...extractFilters(query),
+  };  
+}
+
+function applyDefultsAndAliases(query: Query<Partial<GroupFilters>>): Query<Partial<GroupFilters>> {
```
Comment (eladex) [inline src/group/organizationGroup/organizationGroup.extractQuery.ts:45]: add alias 'all' for isAlive

File: src/person/person.controller.ts [hunk 23]
```diff
@@ -305,7 +321,16 @@ export class Person {
...
-    return esRepository.search(query);
+  static async searchPersonsByName(query: Partial<PersonSearchQuery>) {
+    const { fullName, underGroupId, ...rest } = query;
+    const filters: Partial<PersonTextSearchFilters> = rest;
+    // the query makes sense only if 'fullName' is requested
+    if (!fullName) return [];
+    // get the group to search persons under it's hierarchy path
+    if (!!underGroupId) {
+      const group = await Person._organizationGroupRepository.findById(underGroupId);
+      if (!!group) filters.hierarchyPath = [...group.hierarchy, group.name].join('/');
```
Comment (eladex) [inline src/person/person.controller.ts:332]: throw error if group does not exist (can use controller function)

File: src/person/person.elasticSearchRepository.ts [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+    const query = {
+      fullName,
+      ...filters,
+    };
+    const body = QueryBuilder.buildBoolQuery(query, {
+      fullName: {
+        context: FieldContext.Query,
+        fuzzy: true,
+      },
+    });
```
Comment (eladex) [inline src/person/person.elasticSearchRepository.ts:33]: delete

File: src/person/person.extractQuery.ts [hunk 25]
```diff
@@ -1,46 +1,35 @@
...
-  mobilephone: 'mobilePhone',
-  fullname: 'fullName',
-};
+const { aliases: { persons: personAliases }, defaults: { persons: personDefaults } } = config.queries;
 
-export const filterQueryAllowedFields = ['currentUnit', 'domainUsers.dataSource', 'entityType', 
-  'job', 'rank', 'responsibility', 'serviceType', 'status'];
+const extractFilterKeys: (keyof PersonFilters)[] = ['currentUnit', 'domainUsers.dataSource', 
+  'entityType', 'job', 'rank', 'responsibility', 'serviceType', 'status', 'underGroupId'];
+const extractSearchKeys: (keyof PersonSearchQuery)[] = [...extractFilterKeys, 'fullName'];
```
Comment (eladex) [inline src/person/person.extractQuery.ts:10]: can be just ['fullName'] (need refactor to the extract functions)

File: src/group/organizationGroup/organizationGroup.controller.ts (hunk 19 above)
Comment (eladex) [reply src/group/organizationGroup/organizationGroup.controller.ts:261]: searchByNameAndHierarchy takes care of it

**Review by DavidHeymann — APPROVED** [review]
Finally, shcoyech!!

**Pull request merged**